doc/port_fec-ber/port_fec_ber.md

```diff
@@ -0,0 +1,201 @@
+
+### Definitions / Abbreviations
+
+ | Term    |  Definition / Abbreviation                                            |
+ |---------|-----------------------------------------------------------------------|
+ | FEC     | Forward Error Correction  |
+ | BER     | Bits Error Rate, a measure of percentage of transmitted bits that are received incorrectly.  
+ | Pre FEC | The number of bits which FEC successfully correct.  
+ | Post FEC| The number of bits which FEC fails to correct.  
+ | Frame   | Size of each FEC block.  
+ | Symbol  | Part of the FEC structure which the error detection and correction base on.  
+ | RS-FEC  | Reed Solomon Forward Error correction, RS-544 = 5440 total size , RS-528 = 5280 total size  
+ | NRZ     | Non Return to Zero encoding  
+ | PAM4    | Pulse Amplitude Modulation 4 level encoding  
+
+### 1 Overview
+
+FEC is a common hardware feature deployed in a high speed interconnect. Due to the signal integrity issue, data being transfer might have bit(s) corruption. The FEC will correct the data's corruption and increment counters to account for corrected bits (Pre FEC) or uncorrected frame (Post FEC)
+
+## 2 Requirements
+### 2.1 Functional Requirements
+  This HLD is to   
+  - enhance the current "show interface counter fec-stat" to include Pre and Post BER statistic as new columns
+  - Add Pre and Post FEC BER per interface into Redis DB for telemetry streaming
+  - Calculate the Pre and Post FEC BER at the same interval as the PORT_STAT poll rate which is 1 sec.
+
+### 2.2 CLI Requirements
+
+The existing "show interface counter fec-stat" will be enhanced to include two additional columns for BER.  
+ - Pre FEC BER  
+ - Post FEC BER
+
+## 3 Architecture Design
+
+There are no changes in the current Sonic Architecture.
+
+## 4 High-Level Design
+
+ * SWSS changes:
+   + port_rates.lua
+
+      Enhance to collect and compute the BER on each port at the same port state collection interval. It is currently at 1 second.
+
+     - Access the counter_db for counters for SAI_PORT_STAT_IF_IN_FEC_CORRECTED_BITS & SAI_PORT_STAT_IF_IN_FEC_NOT_CORRECTABLE_FRAMES
+     - Access to the appl_db to compute the actual serdes speed of the ports and its number of lanes
+     - Store the computed BER and the old redis counters value back to the redis DB
+
+ * Utilities Common changes:
+
+   + portstat.py:
+
+     The portstat command with -f , which representing the cli "show interface counter fec-stat" will enhanced to add two new columns, FEC_PRE_BER & FEC_POST_BER
+
+   + netstat.py :
+
+     Add new format support to display the BER in a floating point format
+
+
+### 4.1 Assumptions
+
+SAI provide access to each interface the following attributes
+- SAI_PORT_STAT_IF_IN_FEC_CORRECTED_BITS
+  - monotonically increasing
+  - return not support if its not working for an interface
+- SAI_PORT_STAT_IF_IN_FEC_NOT_CORRECTABLE_FRAMES
+  - monotonically increasing
```

Comment:
after link status change the counter isn't reliable. After offline discussion with Microsoft it was decided to remove this requirement

Reply:
requirement removed from the doc